src/shared/elf-util.c

```diff
@@ -0,0 +1,817 @@
+                        DLSYM_ARG(dwfl_thread_tid),
+                        DLSYM_ARG(dwfl_thread_getframes),
+                        DLSYM_ARG(dwfl_getthreads),
+                        DLSYM_ARG(dwfl_offline_section_address));
+        if (r <= 0)
+                return r;
+
+        return 1;
+}
+
+static int dlopen_elf(void) {
+        int r;
+
+        r = dlopen_many_sym_or_warn(
+                        &elf_dl, "libelf.so.1", LOG_DEBUG,
+                        DLSYM_ARG(elf_begin),
+                        DLSYM_ARG(elf_end),
+                        DLSYM_ARG(elf_getphdrnum),
+                        DLSYM_ARG(elf_getdata_rawchunk),
+                        DLSYM_ARG(elf_errmsg),
+                        DLSYM_ARG(elf_errno),
+                        DLSYM_ARG(elf_memory),
+                        DLSYM_ARG(elf_version),
+                        DLSYM_ARG(gelf_getehdr),
+                        DLSYM_ARG(gelf_getphdr),
+                        DLSYM_ARG(gelf_getnote));
+        if (r <= 0)
+                return r;
+
+        return 1;
+}
+
+struct stack_context {
+        FILE *f;
+        Dwfl *dwfl;
+        Elf *elf;
+        unsigned n_thread;
+        unsigned n_frame;
+        JsonVariant **package_metadata;
+        Set **modules;
+};
+
+static struct stack_context *stack_context_destroy(struct stack_context *c) {
+        if (!c)
+                return NULL;
+
+        if (c->f)
+                c->f = safe_fclose(c->f);
+
+        if (c->dwfl) {
+                sym_dwfl_end(c->dwfl);
+                c->dwfl = NULL;
+        }
+
+        if (c->elf) {
+                sym_elf_end(c->elf);
+                c->elf = NULL;
+        }
+
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct stack_context *, stack_context_destroy);
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(Elf *, sym_elf_end, NULL);
+
+static int frame_callback(Dwfl_Frame *frame, void *userdata) {
+        struct stack_context *c = userdata;
+        Dwarf_Addr pc, pc_adjusted, bias = 0;
+        _cleanup_free_ Dwarf_Die *scopes = NULL;
+        const char *fname = NULL, *symbol = NULL;
+        Dwfl_Module *module;
+        bool is_activation;
+        uint64_t module_offset = 0;
+
+        assert(frame);
+        assert(c);
+
+        if (c->n_frame >= FRAMES_MAX)
+                return DWARF_CB_ABORT;
+
+        if (!sym_dwfl_frame_pc(frame, &pc, &is_activation))
+                return DWARF_CB_ABORT;
+
+        pc_adjusted = pc - (is_activation ? 0 : 1);
+
+        module = sym_dwfl_addrmodule(c->dwfl, pc_adjusted);
+        if (module) {
+                Dwarf_Die *s, *cudie;
+                int n;
+                Dwarf_Addr start;
+
+                cudie = sym_dwfl_module_addrdie(module, pc_adjusted, &bias);
+                if (cudie) {
+                        n = sym_dwarf_getscopes(cudie, pc_adjusted - bias, &scopes);
+                        for (s = scopes; s < scopes + n; s++) {
```

Comment:
This seems to trigger UBSan:
```
../src/shared/elf-util.c:210:53: runtime error: applying zero offset to null pointer
    #0 0x7f0f51aee1c6 in frame_callback /home/vagrant/systemd/build/../src/shared/elf-util.c:210:53
    #1 0x7f0f4ebc0f7a in dwfl_thread_getframes (/lib64/libdw.so.1+0x5ff7a)
    #2 0x7f0f51aedaff in thread_callback /home/vagrant/systemd/build/../src/shared/elf-util.c:262:13
    #3 0x7f0f4ebc0db9 in dwfl_getthreads (/lib64/libdw.so.1+0x5fdb9)
    #4 0x7f0f51ae9dde in parse_core /home/vagrant/systemd/build/../src/shared/elf-util.c:552:13
    #5 0x7f0f51ae8989 in parse_elf /home/vagrant/systemd/build/../src/shared/elf-util.c:639:21
    #6 0x4775e0 in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/fuzz/fuzz-core.c:27:16
    #7 0x4308b9 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/out/fuzz-core+0x4308b9)
    #8 0x431420 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/vagrant/systemd/out/fuzz-core+0x431420)
    #9 0x433969 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-core+0x433969)
    #10 0x433e20 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/vagrant/systemd/out/fuzz-core+0x433e20)
    #11 0x41949a in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/out/fuzz-core+0x41949a)
    #12 0x408ca6 in main (/home/vagrant/systemd/out/fuzz-core+0x408ca6)
    #13 0x7f0f50fe0b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #14 0x408cfd in _start (/home/vagrant/systemd/out/fuzz-core+0x408cfd)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/shared/elf-util.c:210:53 in
```

Reply:
Not addressed yet?

Reply:
it's a pre-existing condition, this was just moved to a new file (and dlopen'ed), nothing else

Reply:
but it's easy enough, so added a commit